File: _includes/sort-benchmark/algorithms/merge.rs
fn merge_sort(a: &mut [usize]) {
    let n = a.len();
    if n <= 1 {
        return;
    }
    let mid = n / 2;
    merge_sort(&mut a[..mid]);
    merge_sort(&mut a[mid..]);
    let mut merged = Vec::with_capacity(n);
    let (mut l, mut r) = (0, mid);
    while l < mid && r < n {
        if a[l] <= a[r] {
            merged.push(a[l]);
            l += 1;
        } else {
            merged.push(a[r]);
            r += 1;
        }
    }
    merged.extend_from_slice(&a[l..mid]);
    merged.extend_from_slice(&a[r..]);
    a.copy_from_slice(&merged);
}

File: _includes/sort-benchmark/algorithms/merge.swift
func merge_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { merge_sort($0) }
}

func merge_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    let n = a.count
    if n <= 1 {
        return
    }
    let mid = n / 2
    merge_sort(UnsafeMutableBufferPointer(rebasing: a[0..<mid]))
    merge_sort(UnsafeMutableBufferPointer(rebasing: a[mid..<n]))
    var merged = [Int]()
    merged.reserveCapacity(n)
    var l = 0
    var r = mid
    while l < mid && r < n {
        if a[l] <= a[r] {
            merged.append(a[l])
            l += 1
        } else {
            merged.append(a[r])
            r += 1
        }
    }
    while l < mid {
        merged.append(a[l])
        l += 1
    }
    while r < n {
        merged.append(a[r])
        r += 1
    }
    for i in 0..<n {
        a[i] = merged[i]
    }
}
